service/matching/tasklist/isolation_balancer.go

```diff
@@ -0,0 +1,422 @@
+	if len(writePartitions) == 1 {
+		// If we resized to 1 partition, remove all assignments from that partition and clear all group status
+		// There's nothing to do while there's only 1
+		if len(writePartitions[0].IsolationGroups) != 0 {
+			i.reset()
+			return map[int]*types.TaskListPartition{
+				0: {},
+			}, true
+		}
+		return writePartitions, false
+	}
+	i.refreshGroups(metrics, writePartitions)
+	partitionCount := i.getPartitionsPerGroup(writePartitions)
+
+	partitionGoal := i.calculatePartitionGoal(metrics, partitionCount, len(writePartitions))
+
+	return i.applyGroupChanges(metrics, partitionCount, partitionGoal, writePartitions)
+}
+
+func (i *isolationBalancer) reset() {
+	maps.Clear(i.groupState)
+}
+
+func (i *isolationBalancer) refreshGroups(metrics *aggregatePartitionMetrics, writePartitions map[int]*types.TaskListPartition) {
+	polledOrAssigned := make(map[string]bool)
+	// Track it for resizing/removal
+	for _, p := range writePartitions {
+		for _, group := range p.IsolationGroups {
+			polledOrAssigned[group] = true
+			if _, ok := i.groupState[group]; !ok {
+				i.groupState[group] = i.createIsolationGroupState(true)
+			}
+		}
+	}
+	// Track it as we might assign it if they maintain pollers
+	for group := range metrics.hasPollersByIsolationGroup {
+		polledOrAssigned[group] = true
+		if _, ok := i.groupState[group]; !ok {
+			i.groupState[group] = i.createIsolationGroupState(false)
+		}
+	}
+	for group := range i.groupState {
+		// No pollers and no write partitions means we don't need to track it
+		if !polledOrAssigned[group] {
+			delete(i.groupState, group)
+			continue
+		}
+	}
+}
+
+func (i *isolationBalancer) createIsolationGroupState(assignToPartitions bool) *isolationGroupState {
+	return &isolationGroupState{
+		underLoad:             clock.NewSustain(i.timeSource, i.config.IsolationGroupDownscaleSustainedDuration),
+		overLoad:              clock.NewSustain(i.timeSource, i.config.IsolationGroupUpscaleSustainedDuration),
+		noPollers:             clock.NewSustain(i.timeSource, i.config.IsolationGroupNoPollersSustainedDuration),
+		hasPollers:            clock.NewSustain(i.timeSource, i.config.IsolationGroupHasPollersSustainedDuration),
+		canAssignToPartitions: assignToPartitions,
+	}
+}
+
+func (i *isolationBalancer) getPartitionsPerGroup(writePartitions map[int]*types.TaskListPartition) map[string]int {
+	partitionCount := make(map[string]int)
+	for _, p := range writePartitions {
+		for _, group := range p.IsolationGroups {
+			partitionCount[group]++
+		}
+	}
+	return partitionCount
+}
+
+func (i *isolationBalancer) calculatePartitionGoal(aggregateMetrics *aggregatePartitionMetrics, partitionCountByGroup map[string]int, writePartitionCount int) map[string]int {
+	partitionGoal := make(map[string]int)
+	assignableGroups := 0
+	// Identify groups that have switched to assignable and count them. This impacts the minimum partitions per group
```

Comment:
under what conditions does this happen?

Reply:
Whether or not an isolation group is assignable is determined by whether it has sustained pollers from that isolation group. If we lose pollers (zonal drain) then we stop assigning that isolation group and allow tasks for it to be evenly distributed across all partitions.